lib/spack/spack/solver/concretize.lp

```diff
@@ -495,6 +495,39 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
 #defined external_spec_condition/4.
 #defined external_spec_condition/5.
 
+%-----------------------------------------------------------------------------
+% Config required semantics
+%-----------------------------------------------------------------------------
+
+% A package may have an associated "requirement_group": a set of specs where
+% at least one must be applied. Each spec in the group is associated
+% with a condition via asp.impose().
+{ condition_holds(Y) : requirement_group_member(Y, Package, X) } :-
+  node(Package),
+  requirement_group(Package, X).
```

Comment:
(bug) Here I think we have a logical error that I need to think how to solve. The problem is that we might fulfill all the conditions to `impose(Y)` without clingo deciding to create a `condition_holds(Y)` atom. This is related to `condition_holds` appearing in a choice rather than a rule. To be more clear, if the requirement is `zlib@1.2.12`, clingo might decide not to have the corresponding `condition_holds` atom in the answer set, but still derive `version(zlib, 1.2.12)` by other means.

Reply:
@scheibelp I have to run unit-tests, but I think this is a possible solution to the bug:
```diff
diff --git a/lib/spack/spack/solver/asp.py b/lib/spack/spack/solver/asp.py
index 31d765feaa..dde7898c3a 100644
--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -945,10 +945,9 @@ def cfg_required_rules(self, pkg):
                 spec = spack.spec.Spec(spec_str)
                 if not spec.name:
                     spec.name = pkg_name
-                member_id = next(self._condition_id_counter)
+                member_id = self.condition(spec, imposed_spec=spec, name=pkg_name)
                 self.gen.fact(fn.requirement_group_member(member_id, pkg_name, cnd_grp_id))
                 self.gen.fact(fn.requirement_has_weight(member_id, requirement_weight))
-                self.impose(member_id, spec, name=pkg_name, node=False)
                 requirement_weight += 1
             cnd_grp_id += 1
 
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 17ace1aa5a..deca651868 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -504,18 +504,10 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
 % Config required semantics
 %-----------------------------------------------------------------------------
 
-% A package may have an associated "requirement_group": a set of specs where
-% at least one must be applied. Each spec in the group is associated
-% with a condition via asp.impose().
-{ condition_holds(Y) : requirement_group_member(Y, Package, X) } :-
-  node(Package),
-  requirement_group(Package, X).
-
 requirement_group_satisfied(Package, X) :-
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) } ,
   node(Package),
-  requirement_group(Package, X),
-  requirement_group_member(Y, Package, X),
-  condition_holds(Y).
+  requirement_group(Package, X).
 
 error(2, "Cannot satisfy requirement group for package '{0}'", Package) :-
   node(Package),

```

Reply:
Not that above you can change the semantics from `any_of` to `one_of` by just adding an upper bound to the cardinality constraint:
```
 1 { condition_holds(Y) : requirement_group_member(Y, Package, X) } 1
```

Reply:
I have run tests succesfully on:
* **Spack:** 0.19.0.dev0 (8577f4c9401f14d6d017eeb936b1b62bc246c63b)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

Reply:
> clingo might decide not to have the corresponding condition_holds atom in the answer set, but still derive version(zlib, 1.2.12) by other means.

That is true, but I considered that fine: if you combine my original constraint with the below constraints that generate error facts, my intent was that an error would be generated if you cannot derive `condition_holds(Y)`.

I originally used cardinality constraints in this rule, but shifted to this approach in light of https://github.com/spack/spack/pull/30669/files#diff-09156cc9ff0b92ff5a0082e65d0c6ca11ecb86b1af9e719ff29d8445a0a8d332.

If we applied your changes, I assume the below rules like 

```
error(2, "Cannot satisfy requirement group for package '{0}'", Package) :-
   node(Package),
   requirement_group(Package, X),
   not requirement_group_satisfied(Package, X).
```

would never be triggered (i.e. the concretizer would just say it couldn't find a solution vs. producing a solution with error facts) - is that true? Or does your approach also make use of the error facts?

> Not that above you can change the semantics from any_of to one_of by just adding an upper bound to the cardinality constraint:

To be clear I think you mean that when you see `one_of` you interpret it to mean "exactly one of" (I didn't interpret it that way, but I can proceed with that interpretation). I don't think "exactly-one-of" semantics are desirable: I think this might drive the concretizer away from sensible solutions that happen to satisfy more than one spec in a requirement group (and I don't think users would ever depend on the mutual exclusion).

Therefore I should rename this to "any_of" (which I originally avoided because in other contexts - e.g. variants - "any of" also allows for choosing none of the options) or perhaps even better: "at_least_one_of" (which I think would eliminate all ambiguities).

Reply:
> is that true? Or does your approach also make use of the error facts?

The approach is only a slight variation of yours and generates error facts. With a "exactly one of" semantic and this:
```yaml
packages:
  zlib:
    require:
    - one_of: ['%gcc', '%gcc']
```
I get:
```console
$ spack spec zlib
==> Error: Cannot satisfy requirement group for package 'zlib'
```
The point is that I use a cardinality constraint in the body of the rule to derive requirement_group_satisfied(Package, X), and I enforce the full condition.

I think the current approach is inconsistent internally. You will not have `condition_holds(Y)` for conditions that actually hold, and the weight being used is that of the first constraint matching. So at least we have a mismatch between the name of the atom and its meaning.

Using `one_of` (in the sense of "exactly one of") can lead to a more straightforward way of dealing with weights, and I think gives the same expressiveness to users.